### ISSUE TITLE
Make more clear which ops are supported by plugins

### ISSUE THREAD

**Issue by co9olguy** [XanaduAI/pennylane#271]
Currently if a user wants to determine if a particular operation is supported on a device from the documentation (e.g., in the [`default.qubit` plugin](https://pennylane.readthedocs.io/en/latest/plugins/default_qubit.html)), they will find a bunch of gates/operations/observables listed, but under different names than found in PennyLane. This may be confusing to users. We should state very clearly in the documentation what operations that device supports, using the native pennylane names.

As well, `default.qubit` docstrings should be updated to reflect the full suite of observables that it actually supports (right now, it only indicates that it supports X, Y, and Z gates).

**Comment by josh146** [comment]
:+1:

> As well, default.qubit docstrings should be updated to reflect the full suite of observables that it actually supports (right now, it only indicates that it supports X, Y, and Z gates).

This definitely needs updating, good catch.

---

Something I've been thinking about for a while; the two docs pages

* https://pennylane.readthedocs.io/en/latest/plugins/default_qubit.html
* https://pennylane.readthedocs.io/en/latest/plugins/default_gaussian.html

are not user facing; there are too many layers of abstraction between the `Device` class and the QNode for any of the documentation to make any sense to users. Instead, it is only relevant to plugin developers as a reference to work on.

@mariaschuld and I had a plan in the table of contents refactor to better represent this; see https://pennylane.readthedocs.io/en/doc-refactor/.

Here, the reference plugins are nested and within the 'Developer API` section of the TOC:

![image](https://user-images.githubusercontent.com/2959003/61766438-6e522680-ae13-11e9-93ff-0aa86f2a3718.png)

---

Currently, the recommend way of determining allowed operations and observables is the `dev.operations` and `dev.observables` required properties of the `Device` class. This must always exist, and by design will _always_ match exactly what is allowed on the device:

```python
>>> import pennylane as qml    
>>> dev = qml.device("default.qubit", wires=2
>>> dev.observables
{'Hadamard', 'Hermitian', 'Identity', 'PauliX', 'PauliY', 'PauliZ'}
```

We should emphasize this in the documentation (in the tutorials etc.), and find a good user facing page in the documentation to list this for the reference plugins.

**Comment by AroosaIjaz** [comment]
> > As well, default.qubit docstrings should be updated to reflect the full suite of observables that it actually supports (right now, it only indicates that it supports X, Y, and Z gates).

I agree! I also noticed this and found it confusing when adding more gates to Pennylane. 
 
> Currently, the recommend way of determining allowed operations and observables is the `dev.operations` and `dev.observables` required properties of the `Device` class. This must always exist, and by design will _always_ match exactly what is allowed on the device:
> 
> ```python
> >>> import pennylane as qml
> >>> dev = qml.device("default.qubit", wires=2)
> >>> dev.observables
> {'Hadamard', 'Hermitian', 'Identity', 'PauliX', 'PauliY', 'PauliZ'}
> ```
> 
> We should emphasize this in the documentation (in the tutorials etc.), and find a good user facing page in the documentation to list this for the reference plugins.

I can add this in the tutorials. 

> Something I've been thinking about for a while; the two docs pages
> 
> * https://pennylane.readthedocs.io/en/latest/plugins/default_qubit.html
> * https://pennylane.readthedocs.io/en/latest/plugins/default_gaussian.html
> 
> are not user facing; there are too many layers of abstraction between the `Device` class and the QNode for any of the documentation to make any sense to users. Instead, it is only relevant to plugin developers as a reference to work on.

I agree. As a new user, I found the `Reference Plugin` section redundant and did not understand that it is meant for developers. 
 
> @mariaschuld and I had a plan in the table of contents refactor to better represent this; see https://pennylane.readthedocs.io/en/doc-refactor/.
>
This looks pretty good :)

**Comment by co9olguy** [comment]
@josh146 I don't think that plan matches the reality. As you know, while the default qubit plugin was meant to be a reference for developers, it actually ends up being used by pretty much every user (partially our fault, since it is used as example in all tutorials). To reflect this reality, we should create a user-focused doc page for this plugin (we can also keep a less obvious developer page)

**Comment by josh146** [comment]
I agree 💯, we should have a page in the docs for using the default qubit (and gaussian) plugin. However, as Aroosa mentioned, it shouldn't be the existing `default_qubit.rst` page that is autogenerated via sphinx.

This docs page will contain all methods of the `DefaultQubitDevice` class, such as `apply`, `pre_measure`, `expval`, that are never accessible by the user, so it is very confusing (hence Maria and I moving it in the ToC refactor).

We had this same issue with the other plugins. The solution there was to have a separate page in the frontmatter (`deviceusage.rst`) that explicitly listed supported operations, options, and usage of the device.

**Comment by josh146** [comment]
I think we're in agreement, the plan I discussed above was just what to do with the existing autogen page, and not the plan for an upcoming user facing device page :laughing:

**Comment by AroosaIjaz** [comment]
I think @josh146 , you know better how to separate the developer part of the current plugin page into a separate one. So, I will only take care of adding something on `dev.operations` and `dev.observables` to the tutorials.

**Comment by mariaschuld** [comment]
Wanting to revive this discussion after the docs refactor. It would be very useful to have a table on the "plugins" [landing page](https://pennylane.ai/plugins.html) listing all core operations against all plugins, with ticks if the operation is supported by the plugin. 

In addition to that, it would be cool to have icons of the supporting plugins in the[ list of operations](https://pennylane.readthedocs.io/en/stable/introduction/operations.html) in the docs, but that seems harder. 

But overall, this should be updated automatically. From a place where all plugins are installed one could call ``dev.operations`` and ``dev.observables`` and auto-generate a table. Does anyone have an idea how to do this?

**Comment by johannesjmeyer** [comment]
I guess with the `decomposition` feature, this is nothing people have to worry about anymore, don't they?

**Comment by josh146** [comment]
Yep. Not only that, but the new QubitDevice brings support to all observables as well.

**Comment by yishairasowsky** [comment]
did i understand correctly that this issue is moot. hmm, i wonder then why it isn't closed. i am a new participant here. just hoping to contribute something to your great work.

**Comment by mariaschuld** [comment]
Maybe we should still make a note on the Plugins page to state how people should understand which gates they can use? Also, what about plugin custom gates, like ``cirq``'s  CCNOT? So far a user would have no chance to know about it, right?

**Comment by josh146** [comment]
> Also, what about plugin custom gates, like cirq's CCNOT? So far a user would have no chance to know about it, right?

This is already available as `qml.Toffoli` in PennyLane; we should remove the custom `CCNOT` gate from `cirq`.

Although, now that I think about it, we should make `qml.CCNOT` available as a 'alias' to `qml.Toffoli`, since someone might know one name but not the other!

In general, we should try and remove custom plugin ops as much as possible, and have everything accessible in PennyLane core.

**Comment by co9olguy** [comment]
@yishairasowsky @mariaschuld @josh146 I think the original intention of this issue was to make it clear to people, with a quick glance of a plugin's docs, which operations that plugin supports (or doesn't). Digressions about individual gates (like `CCNOT` aside), is it currently true that all (qubit-model) plugins support all PennyLane core gates, and there are no gates which are supported by some plugins which are not in pennylane core?

**Comment by co9olguy** [comment]
@yishairasowsky A really valuable contribution could be to add additional gates to the PennyLane codebase which are not currently available. Part of that contribution could be to determine which gates we should add (e.g., gates that are appearing on other platforms or hardware)

**Comment by mridulsar** [comment]
I was recently perusing through Q# and noticed they had two nifty features: Block() which takes quantum gates and 'strings' them together, through tensor/matrix calculations. As well as ApplyToEach() which applies any gate across each qubit in your system.
Though they aren't gates, they make them much easier to work with! especially over large amounts of qubits. 
As far as gates I have been interested to find a sort of template gate for Conditional Phase Shift Transform. I am a bit of a newcomer on this particular gate so if it seems far fetched you are probably right.
Best
Mridul Sarkar

**Comment by josh146** [comment]
Hi @mridulsar! The first two operations (meta-operations?) sounds like our [`qml.broadcast()`](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.broadcast.html) function. These allow you to take any combination of gates, and apply them in a 'pattern' across multiple qubits in one go 🙂

**Comment by mridulsar** [comment]
Hello @josh146 ! I was using broadcast just yesterday and hit a bug when trying to call Hadmard within Broadcast. The debugger let me know Hadmard object is not callable, I assumed from inside broadcoast. Do you know of a way to circumvent this? Should I be using a different notation when calling gates within Broadcast?

**Comment by josh146** [comment]
@mridulsar I'll have to get you to post a small code example so I can see the relevant code/reproduce it myself. The best place to do this is on the [Xanadu PennyLane Forum](https://discuss.pennylane.ai)
